sdk/eventhub/azure-eventhub/azure/eventhub/_producer_client.py

```diff
@@ -232,12 +451,64 @@ def from_connection_string(cls, conn_str, **kwargs):
+        :keyword float timeout: The maximum wait time to send the event data.
+         If not specified, the default wait time specified when the producer was created will be used.
+        :keyword str partition_id: The specific partition ID to send to. Default is None, in which case the service
+         will assign to all partitions using round-robin.
+         A `TypeError` will be raised if partition_id is specified and event_data_batch is an `EventDataBatch` because
+         `EventDataBatch` itself has partition_id.
+        :keyword str partition_key: With the given partition_key, event data will be sent to
+         a particular partition of the Event Hub decided by the service.
+         A `TypeError` will be raised if partition_key is specified and event_data_batch is an `EventDataBatch` because
+         `EventDataBatch` itself has partition_key.
+         If both partition_id and partition_key are provided, the partition_id will take precedence.
+         **WARNING: Setting partition_key of non-string value on the events to be sent is discouraged
+         as the partition_key will be ignored by the Event Hub service and events will be assigned
+         to all partitions using round-robin. Furthermore, there are SDKs for consuming events which expect
+         partition_key to only be string type, they might fail to parse the non-string value.**
+        """
+        partition_id = kwargs.get("partition_id") or ALL_PARTITIONS
+        partition_key = kwargs.get("partition_key")
+        send_timeout = kwargs.get("timeout")
+        if partition_key:
+            set_message_partition_key(event_data.message, partition_key)
+        try:
+            try:
+                cast(EventHubProducer, self._producers[partition_id]).send(
+                    event_data, timeout=send_timeout
+                )
+            except (KeyError, AttributeError, EventHubError):
+                self._start_producer(partition_id, send_timeout)
+                cast(EventHubProducer, self._producers[partition_id]).send(
+                    event_data, timeout=send_timeout
+                )
+            if self._on_success:
+                self._on_success(event_data, partition_id)
+        except Exception as exc:
+            if self._on_error:
+                self._on_error(event_data, exc, partition_id)
+            else:
+                raise
+
     def send_batch(self, event_data_batch, **kwargs):
         # type: (Union[EventDataBatch, SendEventTypes], Any) -> None
-        """Sends event data and blocks until acknowledgement is received or operation times out.
+        """
+        Sends a batch of event data.
+        By default the method will blocks until acknowledgement is received or operation times out.
+        If the `EventHubProducerClient` is configured to run in buffered mode, the method will enqueue the events
+        into local buffer and return. The producer will do automatic sending in the background.
```

Comment:
Maybe we should add a comment that in buffered mode, sending a batch will remain intact, and not be 'repackaged'. This may have implications to the efficiency of sending events.

Reply:
comment added, thanks for the suggestion.